### PR TITLE
JAMES-3434 EmailSubmission/set create

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+         |         }
+         |    }
+         |  }, "c1"]]
+         |}""".stripMargin
+
+    `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(requestBob)
+    .when
+      .post.prettyPeek
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:104]: debug :-p

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailSubmissionSetSerializer.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  private implicit val mapCreationRequestByEmailSubmissionCreationId: Reads[Map[EmailSubmissionCreationId, JsObject]] =
+    readMapEntry[EmailSubmissionCreationId, JsObject](s => refineV[NonEmpty](s),
+      {
+        case o: JsObject => JsSuccess(o)
+        case _ => JsError("Expecting a JsObject as a creation entry")
+      })
+
+  private implicit val emailSubmissionSetRequestReads: Reads[EmailSubmissionSetRequest] = Json.reads[EmailSubmissionSetRequest]
+
+  private implicit val emailSubmissionIdWrites: Writes[EmailSubmissionId] = Json.valueWrites[EmailSubmissionId]
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailSubmissionSetSerializer.scala:38]: You could write `type EmailSubmissionId = Id` on the type definition, it enables having ser/deser implicit, and it allows to simplify wiring in other places

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+case class EmailSubmissionSetResponse(accountId: AccountId,
+                                      newState: State,
+                                      created: Option[Map[EmailSubmissionCreationId, EmailSubmissionCreationResponse]])
+
+case class EmailSubmissionId(value: String)
+
+case class EmailSubmissionCreationResponse(id: EmailSubmissionId)
+
+case class Parameters(value: String)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala:49]: Likely not.  The type in the spec is `MAp[String, String]`

```
parameters: Object|null Any parameters to send with the email address (either mail-parameter or rcpt-parameter as appropriate, as specified in [@!RFC5321]). If supplied, each key in the object is a parameter name, and the value is either the parameter value (type String) or null if the parameter does not take a value. For both name and value, any xtext or unitext encodings are removed (see [@!RFC3461] and [@!RFC6533]) and JSON string encoding is applied.
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  override val requiredCapabilities: Capabilities = Capabilities(CORE_CAPABILITY, MAIL_CAPABILITY)
+
+  sealed trait CreationResult {
+    def emailSubmissionCreationId: EmailSubmissionCreationId
+  }
+  case class CreationSuccess(emailSubmissionCreationId: EmailSubmissionCreationId, emailSubmissionCreationResponse: EmailSubmissionCreationResponse) extends CreationResult
+  case class CreationFailure(emailSubmissionCreationId: EmailSubmissionCreationId, exception: Exception) extends CreationResult
+  case class CreationResults(created: Seq[CreationResult]) {
+    def retrieveCreated: Map[EmailSubmissionCreationId, EmailSubmissionCreationResponse] = created
+      .flatMap(result => result match {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:56]:
```
.flatMap {
 case xxx => yyy
}
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        case success: CreationSuccess => Some(success.emailSubmissionCreationId, success.emailSubmissionCreationResponse)
+        case _ => None
+      })
+      .toMap
+      .map(creation => (creation._1, creation._2))
+  }
+
+    override def doProcess(capabilities: Set[CapabilityIdentifier], invocation: InvocationWithContext, mailboxSession: MailboxSession, request: EmailSubmissionSetRequest): SMono[InvocationWithContext] = {
+    for {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:65]: Ident (1 more level?)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        (acc : (CreationResults, ProcessingContext), elem: (EmailSubmissionCreationId, JsObject)) => {
+          val (emailSubmissionCreationId, jsObject) = elem
+          val (creationResult, updatedProcessingContext) = createMailbox(session, emailSubmissionCreationId, jsObject, acc._2)
+          (CreationResults(acc._1.created :+ creationResult), updatedProcessingContext)
+        }
+      }
+      .subscribeOn(Schedulers.elastic())
+  }
+
+  private def createMailbox(mailboxSession: MailboxSession,
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:104]: oups ;-)

**Comment by Arsnael** [comment]
```
[c5815a02092efadcf70b71a291032f228c372879] [INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[c5815a02092efadcf70b71a291032f228c372879] [ERROR] src/main/java/org/apache/james/jmap/rfc8621/RFC8621MethodsModule.java:[28] (imports) ImportOrder: Wrong order for 'org.apache.commons.configuration2.Configuration' import.
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 7]
```diff
@@ -33,31 +33,17 @@ import org.apache.james.jmap.http.UserCredential
...
-   - extra properties are rejected
-   - message not founds are handled
-   - I can chain Email/set create & EmailSubmission/create
-   - multiple recipients
-   - cannot send other people mail
-   - can send delegated emails (read permission)
- */
-
 trait EmailSubmissionSetMethodContract {
+  private def andreAccountId: String = "1e8584548eca20f26faf6becc1704a0f352839f12c208a47fbd486d60f491f7c"
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:46]: Maybe should put in Fixture?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 8]
```diff
@@ -160,6 +146,264 @@ trait EmailSubmissionSetMethodContract {
...
+        .body
+        .asString
+      val responseAndre = `given`(
+        baseRequestSpecBuilder(server)
+          .setAuth(authScheme(UserCredential(ANDRE, ANDRE_PASSWORD)))
+          .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+          .setBody(requestReadMailAndre)
+          .build, new ResponseSpecBuilder().build)
+        .post
+        .`then`
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:315]: indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 9]
```diff
@@ -213,4 +457,512 @@ trait EmailSubmissionSetMethodContract {
...
+         |      "filter": {"inMailbox": "${bobInboxId.serialize}"}
+         |    },
+         |    "c1"]]
+         |}""".stripMargin
+
+    awaitAtMostTenSeconds.untilAsserted { () =>
+      val response = `given`
+        .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+        .body(requestReadMail)
+        .when
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:724]: indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 10]
```diff
@@ -213,4 +457,512 @@ trait EmailSubmissionSetMethodContract {
...
+        .post
+        .`then`
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:726]: indent

**Comment by Arsnael** [comment]
```
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   KeyspaceCreationTest$CreateWhenKeyspaceExists.startShouldNotThrowWhenKeyspaceExists:50 
[696f6825f562df0e222931a668de153e90fc2168] Expecting code not to raise a throwable but caught
[696f6825f562df0e222931a668de153e90fc2168]   <"com.google.inject.CreationException: Unable to create injector, see the following errors:
[696f6825f562df0e222931a668de153e90fc2168] 
[696f6825f562df0e222931a668de153e90fc2168] 1) No implementation for org.apache.james.queue.api.MailQueueFactory<? extends org.apache.james.queue.api.MailQueue> was bound.
[696f6825f562df0e222931a668de153e90fc2168]   while locating org.apache.james.queue.api.MailQueueFactory<? extends org.apache.james.queue.api.MailQueue>
[696f6825f562df0e222931a668de153e90fc2168]     for the 3rd parameter of org.apache.james.jmap.method.EmailSubmissionSetMethod.<init>(EmailSubmissionSetMethod.scala:66)
[696f6825f562df0e222931a668de153e90fc2168]   at org.apache.james.jmap.rfc8621.RFC8621MethodsModule.configure(RFC8621MethodsModule.java:74) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.protocols.JMAPServerModule -> org.apache.james.jmap.draft.JMAPModule -> org.apache.james.jmap.rfc8621.RFC8621MethodsModule)

[...]

[696f6825f562df0e222931a668de153e90fc2168] [ERROR] Errors: 
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   CassandraJamesServerTest » Creation Unable to create injector, see the followi...
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   CassandraNodeConfTest$NormalBehaviour.serverShouldStartServiceWhenNodeIsReachable:78 » Creation
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   CassandraNodeConfTest$OneFailedNode.serverShouldStartServiceWhenNodeIsReachable:99 » Creation
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   CassandraNodeConfTest$UseMappedPort.configShouldWorkWithNonDefaultPort:117 » Creation
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   CassandraWithTikaTest » Creation Unable to create injector, see the following ...
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenImapCommandsReceived » Retryable
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenJmapRequestsReceived » Retryable
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldRetryToConnectToCassandraWhenStartService » Retryable
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldRetryToConnectToElasticSearchWhenStartService:93->assertThatServerStartCorrectly:108 » Creation
[696f6825f562df0e222931a668de153e90fc2168] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldStartAtDefault:75->assertThatServerStartCorrectly:108 » Creation
```
on `james-server-cassandra-guice` module... there is a lot of similar issues

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
There is still a binding missing issue with cassandra guice (not cassandra rabbitmq)... been trying to figure it out but struggling so far

**Comment by Arsnael** [comment]
```
[b40a4c00d6d983f26f6ef77671388ff29caa6944] [ERROR]   CassandraMailboxMapperTest$ConsistencyTest.createAfterPreviousDeleteOnFailedCreateShouldCreateAMailbox:604 Multiple Failures (1 failure)
[b40a4c00d6d983f26f6ef77671388ff29caa6944] 	
[b40a4c00d6d983f26f6ef77671388ff29caa6944] Expected size:<1> but was:<2> in:
[b40a4c00d6d983f26f6ef77671388ff29caa6944] <[Mailbox{id=1b213ce0-1fef-11eb-8426-bfb13ef228f2, namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=INBOX},
[b40a4c00d6d983f26f6ef77671388ff29caa6944]     Mailbox{id=18f3fca0-1fef-11eb-8426-bfb13ef228f2, namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=name}]>
[b40a4c00d6d983f26f6ef77671388ff29caa6944] at CassandraMailboxMapperTest$ConsistencyTest.lambda$createAfterPreviousDeleteOnFailedCreateShouldCreateAMailbox$34(CassandraMailboxMapperTest$ConsistencyTest.java:615)
```
test this please

**Comment by Arsnael** [comment]
:+1:

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+    def retrieveErrors: Map[EmailSubmissionCreationId, SetError] = created
+      .flatMap {
+        case failure: CreationFailure => Some(failure.emailSubmissionCreationId, failure.asSetError)
+        case _ => None
+      }
+      .toMap
+  }
+
+  def init(): Unit = {
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:115]: bracket needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    queue = mailQueueFactory.createQueue(SPOOL)
+  }
+
+  @PreDestroy def dispose(): Unit = {
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:119]: bracket needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+  }
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): SMono[EmailSubmissionSetRequest] =
+    asEmailSubmissionSetRequest(invocation.arguments)
+
+  private def asEmailSubmissionSetRequest(arguments: Arguments): SMono[EmailSubmissionSetRequest] =
+    serializer.deserializeEmailSubmissionSetRequest(arguments.value) match {
+      case JsSuccess(emailSubmissionSetRequest, _) => SMono.just(emailSubmissionSetRequest)
+      case errors: JsError => SMono.raiseError(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:147]: the private method only used once? May be the logic can be written in getRequest() directly?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+  private def create(request: EmailSubmissionSetRequest,
+                     session: MailboxSession,
+                     processingContext: ProcessingContext): SMono[(CreationResults, ProcessingContext)] = {
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:151]: bracket needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+          (CreationResults(acc._1.created :+ creationResult), updatedProcessingContext)
+        }
+      }
+      .subscribeOn(Schedulers.elastic())
+  }
+
+  private def createSubmission(mailboxSession: MailboxSession,
+                            emailSubmissionCreationId: EmailSubmissionCreationId,
+                            jsObject: JsObject,
+                            processingContext: ProcessingContext): (CreationResult, ProcessingContext) = {
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:168]: bracket needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+      case errors: JsError => SMono.raiseError(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  private def create(request: EmailSubmissionSetRequest,
+                     session: MailboxSession,
+                     processingContext: ProcessingContext): SMono[(CreationResults, ProcessingContext)] = {
+    SFlux.fromIterable(request.create
+      .getOrElse(Map.empty)
+      .view)
+      .foldLeft((CreationResults(Nil), processingContext)){
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:155]: space before bracket

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+    } else if (!envelope.mailFrom.email.asString.equals(session.getUser.asString())) {
+      Failure(ForbiddenFromException(envelope.mailFrom.email.asString))
+    } else {
+      Success(mimeMessage)
+    }
+  }
+
+  private def recordCreationIdInProcessingContext(emailSubmissionCreationId: EmailSubmissionCreationId,
+                                                  processingContext: ProcessingContext,
+                                                  emailSubmissionId: EmailSubmissionId) = {
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:248]: bracket needed?

**Review by chibenwa — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**